### PR TITLE
player::takeoff rewrite to use item_location

### PR DESCRIPTION
**Pull request by topherclay**
#### Summary

Infrastructure "rewriting player::takeoff to use item_location instead of item."

#### Purpose of change

I'm rewriting this function in order to give it access to `item_location::obtain` which will help consolidate where the payment of moves happens. 

Currently, 
* some function calls to `takeoff` deduct their own costs before the call, and then the `takeoff` function deducts its cost as a double dip. 
* some function calls leave the move cost payment to the `takeoff` function alone, allowing an unintended discount. 
* some other actions remove worn items without calling `takeoff`, doing their own thing with their own move costs. 

This change will allow the first two cases to absorb the external costs and leave all movement costs for removing worn items to occur inside of `player::takeoff`.

This is also a necessary step for eventually closing #46935. 

#### Describe the solution

I added a temporary overload for `bool player::takeoff( item &it, std::list<item> *res )` that can recieve `item_location` as an argument instead of `item`. 

I looked at each call to the original function and switched it where I could to a call with an `item_location`.

I found three calls which I havent changed, which don't seem to have `item_location` in their scope. I stopped here. 

If anyone knows of a way that I can reconstruct a working `item_location` in place near the calls where only `item` is available, I am curious to see how that works. 

Player::takeoff calls which have not yet been rewritten to use item_location:

* npc.cpp @ 1115  			`took_off = takeoff( *iter );`
* armor_layers.cpp @ 870 	`takeoff( *tmp_worn[leftListIndex] );`
* npcmove.cpp @ 4596		`if( ( needs_change && change_side( elem ) ) || takeoff( elem ) )`

#### Describe alternatives you've considered

Maybe change nothing here, and instead of trying to make `takeoff` consistently determine move costs, let all the individual calls everywhere else determine their own move costs.  

If it's not possible to get `item_location` information in all of these places then it might be a solution to keep the temporary overload and let the temporary overload deduct the move costs. 

#### Testing

For the actions where function calls could successfully be changed to send `item_location` instead of `item`, I ran one character and tried each action to make sure my temporary overload worked as intended.

### CODE REVIEW

**Review by topherclay — COMMENTED** [review]
I found out how to make my own item_locations! 

It seems a little crazy to be created the item_location on the fly here just to feed it into a function that used to be working fine without it...

but because the `obtain` function is on `item_location` and not on `item`, it wasn't actually working fine without it. 

There are two more places in the code where player::takeoff is still called with items instead of item_locations. I will see if this same fix works on those but because they are in npc.cpp and npcmove.cpp, I will not be able to test if my changes work until after I dive into the unit tests and maybe learn how to write my own.  

Edit:
I'm suddenly realizing that the way I'm making item_locations is with the original item that the current code already provides, and with the `character` that I'm getting from `player`. And since the `takeoff` function is already inside of `player` why don't I just make the item_location inside of `takeoff` instead of worrying about how it is called from outside.... 🤦‍♂️

File: src/handle_action.cpp [hunk 1]
```diff
@@ -1273,7 +1273,8 @@ static void takeoff()
     item_location loc = game_menus::inv::take_off( player_character );
 
     if( loc ) {
-        player_character.takeoff( *loc.obtain( player_character ) );
+        loc.obtain( player_character );
```
Comment (Qrox) [inline src/handle_action.cpp:1276]: `obtain` may move the item so I think the returned `item_location` should be passed to `takeoff` here?

File: src/npc.cpp [hunk 2]
```diff
@@ -1112,7 +1112,7 @@ bool npc::wear_if_wanted( const item &it, std::string &reason )
                 return armor.covers( bp );
             } );
             if( iter != worn.end() && !( is_limb_broken( bp ) && iter->has_flag( flag_SPLINT ) ) ) {
-                took_off = takeoff( *iter );
+                took_off = takeoff( *iter ); //Can I turn this *iter into an item_location?
```
Comment (Qrox) [inline src/npc.cpp:1115]: Since `iter` is always from `worn`, I think it's fine to just create the item location as `item_location( *this, &*iter )`.

File: src/npcmove.cpp [hunk 3]
```diff
@@ -4593,7 +4593,8 @@ bool npc::adjust_worn()
         if( !covers_broken( elem, elem.get_side() ) ) {
             const bool needs_change = covers_broken( elem, opposite_side( elem.get_side() ) );
             // Try to change side (if it makes sense), or take off.
-            if( ( needs_change && change_side( elem ) ) || takeoff( elem ) ) {
+            if( ( needs_change && change_side( elem ) ) ||
+                takeoff( elem ) ) { //Can I turn this elem into an item_location?
```
Comment (Qrox) [inline src/npcmove.cpp:4597]: Same here.

File: src/npc.cpp (hunk 2 above)
Comment (topherclay) [reply src/npc.cpp:1115]: Thank you for looking at this. I didn't know you could use both & and * together like that and knowing that helps me not do extra unnecessary steps. Now that I know how easy it is to make an `item_location` object I am rethinking my approach here.

File: src/handle_action.cpp (hunk 1 above)
Comment (topherclay) [reply src/handle_action.cpp:1276]: Thanks, I am getting confused on which changes to item_location are made in place and which you have to catch in the return.

File: src/player.cpp [hunk 4]
```diff
@@ -1942,8 +1942,11 @@ ret_val<bool> player::can_takeoff( const item &it, const std::list<item> *res )
     return ret_val<bool>::make_success();
 }
 
-bool player::takeoff( item &it, std::list<item> *res )
+bool player::takeoff( item_location loc, std::list<item> *res )
 {
+    item it = *loc;
```
Comment (BrettDong) [inline src/player.cpp:1947]: As discussed,
```suggestion
    item &it = *loc;
```

Comment (topherclay) [reply src/player.cpp]: Thank you very much.

**Comment by BrettDong** [comment]
`clang++-9,address,undefined` caught an error in `player::get_dodge with effects` unit test. You may need to update something inside function `float dodge_wearing_item( avatar &dummy, item &clothing )` in `tests/melee_dodge_hit_test.cpp`.

**Comment by topherclay** [comment]
> 
> 
> `clang++-9,address,undefined` caught an error in `player::get_dodge with effects` unit test. You may need to update something inside function `float dodge_wearing_item( avatar &dummy, item &clothing )` in `tests/melee_dodge_hit_test.cpp`.

Thank you @BrettDong, I am slowly learning how to read the important parts out of those logs on the various tests.

File: tests/melee_dodge_hit_test.cpp [hunk 5]
```diff
@@ -55,8 +55,10 @@ static float dodge_with_effect( Creature &critter, const std::string &effect_nam
 static float dodge_wearing_item( avatar &dummy, item &clothing )
 {
     // Get nekkid and wear just this one item
-    std::list<item> temp;
-    while( dummy.takeoff( dummy.i_at( -2 ), &temp ) ) {}
+
+    item_location loc = item_location( *dummy.as_character(), &dummy.i_at( -2 ) );
+    while( dummy.takeoff( loc ) ) {}
```
Comment (Qrox) [inline tests/melee_dodge_hit_test.cpp:60]: I think your change does nothing wrong per se; it's only that the change to use `item_location` started to catch the null pointer returned by `item_location::target`.

I suggest changing the above code to the following to see if it passes the test.

```suggestion
    while( true ) {
        item &it = dummy.i_at( -2 );
        if( it.is_null() ) {
            break;
        }
        item_location loc = item_location( *dummy, &it );
        if( !dummy.takeoff( loc ) ) {
            break;
        }
    }
```

Additionally we may want to add more null pointer checks in the `item_location` code, but that should probably better go in another PR.

File: tests/melee_dodge_hit_test.cpp [hunk 6]
```diff
@@ -56,8 +56,18 @@ static float dodge_wearing_item( avatar &dummy, item &clothing )
 {
     // Get nekkid and wear just this one item
 
-    item_location loc = item_location( *dummy.as_character(), &dummy.i_at( -2 ) );
-    while( dummy.takeoff( loc ) ) {}
+    while( true ) {
+        item &it = dummy.i_at( -2 );
+
+        //If the item is a null pointer then we don't need to take it off.
```
Comment (Qrox) [inline tests/melee_dodge_hit_test.cpp:62]: Strictly speaking this does not check if it's a null pointer, but checks if it's a special `null` item. The function itself should never return a null reference.

Comment (topherclay) [reply tests/melee_dodge_hit_test.cpp:62]: Aw that makes sense. That's funny that even when I take your solution I still add error in the form of a comment!

Thanks for looking at this with me.

File: tests/melee_dodge_hit_test.cpp [hunk 7]
```diff
@@ -55,8 +55,20 @@ static float dodge_with_effect( Creature &critter, const std::string &effect_nam
...
+        if( it.is_null() ) {
+            break;
+        }
+
+        if( !dummy.takeoff( item_location( *dummy.as_character(), &it ) ) ) {
+            break;
+        }
+    }
```
Comment (Qrox) [inline tests/melee_dodge_hit_test.cpp:70]:
```suggestion
    std::list<item> temp;
    while( true ) {
        item &it = dummy.i_at( -2 );

        //If the item is a null item then we don't need to take it off.
        if( it.is_null() ) {
            break;
        }

        if( !dummy.takeoff( item_location( *dummy.as_character(), &it ), &temp ) ) {
            break;
        }
    }
```
If `temp` is not provided, `takeoff` ask the player to handle the item, which calls the input code which leads to the error in the test.

Comment (topherclay) [reply tests/melee_dodge_hit_test.cpp]: Oh wow, that makes sense now why it was named `temp`. Do you know why that item list is named `res` inside `player::takeoff()`?

Comment (Qrox) [reply tests/melee_dodge_hit_test.cpp]: I think it's short for `result`? It returns the items instead of adding them to the player's inventory.

**Comment by topherclay** [comment]
Closed and opened the PR to ping the tests, to see if Travis CI will pass now.
